src/pygeoprocessing/routing/routing.pyx

```diff
@@ -4076,6 +4076,111 @@ def calculate_subwatershed_boundary(
+
+    if raster_info['projection_wkt']:
+        raster_srs = osr.SpatialReference()
+        raster_srs.ImportFromWkt(raster_info['projection_wkt'])
+    else:
+        raster_srs = None
+
+    gpkg_driver = gdal.GetDriverByName('GPKG')
+
+    if os.path.exists(target_outlet_vector_path):
+        LOGGER.warning(
+            f'outlet detection: {target_outlet_vector_path} exists, '
+            'removing before creating a new one.')
+        os.remove(target_outlet_vector_path)
+    outlet_vector = gpkg_driver.Create(
+        target_outlet_vector_path, 0, 0, 0, gdal.GDT_Unknown)
+    outet_basename = os.path.basename(
+        os.path.splitext(target_outlet_vector_path)[0])
+    outlet_layer = outlet_vector.CreateLayer(
+        outet_basename, raster_srs, ogr.wkbPoint)
+    outlet_layer.CreateField(ogr.FieldDefn('i', ogr.OFTInteger))
+    outlet_layer.CreateField(ogr.FieldDefn('j', ogr.OFTInteger))
+    outlet_layer.StartTransaction()
+
+    cdef time_t last_log_time = ctime(NULL)
+
+    cdef _ManagedRaster d8_flow_dir_mr = _ManagedRaster(
+        d8_flow_dir_raster_path_band[0],
+        d8_flow_dir_raster_path_band[1], 0)
+
+    for block_offsets in pygeoprocessing.iterblocks(
+            d8_flow_dir_raster_path_band, offset_only=True):
+        xoff = block_offsets['xoff']
+        yoff = block_offsets['yoff']
+        win_xsize = block_offsets['win_xsize']
+        win_ysize = block_offsets['win_ysize']
+        for yi in range(win_ysize):
+            yi_root = yi+yoff
+            if ctime(NULL) - last_log_time > 5.0:
+                last_log_time = ctime(NULL)
+                current_pixel = xoff + yoff * raster_x_size
+                LOGGER.info(
+                    f'''outlet detection: {
+                        100.0 * current_pixel / <float>(
+                            raster_x_size * raster_y_size):.1f} complete''')
+            for xi in range(win_xsize):
+                xi_root = xi+xoff
+                flow_dir = <int>d8_flow_dir_mr.get(xi_root, yi_root)
+                if flow_dir == flow_dir_nodata:
+                    continue
+
+                xi_n = xi_root+D8_XOFFSET[flow_dir]
+                yi_n = yi_root+D8_YOFFSET[flow_dir]
+                if (xi_n < 0 or xi_n >= raster_x_size or
+                        yi_n < 0 or yi_n >= raster_y_size) or (
+                        <int>d8_flow_dir_mr.get(xi_n, yi_n) ==
+                        flow_dir_nodata):
+                    outlet_point = ogr.Geometry(ogr.wkbPoint)
+                    proj_x, proj_y = gdal.ApplyGeoTransform(
+                        raster_info['geotransform'],
+                        xi_root+0.5, yi_root+0.5)
+                    outlet_point.AddPoint(proj_x, proj_y)
+
+                    outlet_feature = ogr.Feature(outlet_layer.GetLayerDefn())
+                    outlet_feature.SetGeometry(outlet_point)
+                    outlet_feature.SetField('i', xi_root)
+                    outlet_feature.SetField('j', yi_root)

```

Comment:
I'm wondering about including `i` and `j` as the original raster coordinates, what would be the use case for that? It seems potentially confusing to give each point an extra pair of coordinates. And without additional metadata like the pixel size and original spatial reference, would `i` and `j` lose their meaning?

Reply:
I mentioned it in the docstring and added a comment in c134127e48167e25e946523b4604cd9e05a50319 explaining this.. these aren't the projected coordinates but the raster i,j coordinates. It's useful for when you're using this datastructure to seed an upstream search from a pixel coordinate rather than taking the step of having the projected coordinate and then inverse geotransforming it back into raster space. All the hydrological algorithms we have operate at the raster/pixel scale so that's just a handy way of getting it back down there.

A very concrete example of this is where I just used it where I needed to calculate the number of beneficiaries "downstream" from a point. You can do this by using a "distance to channel" function with a weighted raster that is the beneficiary. However you need to also provide where your channels are. Since I actually wanted it to go to the outlet I use this data structure to make a raster that only has the pixels set to 1 where these points are.

I see what you mean that it looks extra and could be derived anyway, but the use case for this is in the raster space not the projected space. In that context the extra information is actually the projected coordinates, but it's nice to have those because then you can just plink it down in qgis and everything matches up.

How does that sound?

Reply:
Makes sense! thanks for the explanation